tree_assume/src/interpreter.rs

```diff
@@ -0,0 +1,408 @@
+                    Order::Sequential | Order::Parallel => {
+                        (self.interpret_expr(e1, arg), self.interpret_expr(e2, arg))
+                    }
+                    Order::Reversed => {
+                        let v2 = self.interpret_expr(e2, arg);
+                        let v1 = self.interpret_expr(e1, arg);
+                        (v1, v2)
+                    }
+                };
+                let Tuple(mut v1) = v1_tuple else {
+                    panic!("expected tuple in extend's first argument in: {:?}", e1)
+                };
+                let Tuple(v2) = v2_tuple else {
+                    panic!("expected tuple in extend's second argument in {:?}", e2)
+                };
+                v1.extend(v2);
+                Tuple(v1)
+            }
+            Expr::Switch(pred, branches) => {
+                let Const(Constant::Int(index)) = self.interpret_expr(pred, arg) else {
+                    panic!("expected integer in switch")
+                };
+                if index < 0 || index as usize >= branches.len() {
+                    // TODO refactor to return a Result
+                    panic!("switch index out of bounds")
+                }
+                self.interpret_expr(&branches[index as usize], arg)
+            }
+            Expr::If(pred, then, els) => {
+                let Const(Constant::Bool(pred_evaluated)) = self.interpret_expr(pred, arg) else {
+                    panic!("expected boolean in if")
+                };
+                if pred_evaluated {
+                    self.interpret_expr(then, arg)
+                } else {
+                    self.interpret_expr(els, arg)
+                }
+            }
+            Expr::DoWhile(input, pred_output) => {
+                let Tuple(mut vals) = self.interpret_expr(input, arg) else {
+                    panic!("expected tuple for input in do-while")
+                };
+                eprintln!("do-while input: {:?}", vals);
+                let mut pred = Const(Constant::Bool(true));
+                while pred == Const(Constant::Bool(true)) {
+                    let Tuple(pred_output_val) =
+                        self.interpret_expr(pred_output, &Some(Tuple(vals.clone())))
+                    else {
+                        panic!("expected tuple for pred_output in do-while")
+                    };
+                    assert!(
+                        pred_output_val.len() == 1 + vals.len(),
+                        "expected pred_output to have one more element than input in {:?}",
+                        pred_output
+                    );
+                    pred = pred_output_val[0].clone();
+                    vals = pred_output_val[1..].to_vec();
+                }
+                eprintln!("do-while result: {:?}", vals);
+                Tuple(vals)
+            }
+            Expr::Let(input, output) => {
+                let vals = self.interpret_expr(input, arg);
```

Comment:
clarify that we evaluate the input to the let here (once), and then bind it to (Arg) when we evaluate the body. ik that's exactly what the code does, but a sentence would help too i think